retry login if IMAP Servers timeout.

hefee opened the pull request:
When connection fails in a recoverable way, we should retry to connect
to the server again. This should help in cases, when the connection to the IMAP server is not 100% stable.

In #7402 it was suggested to implement this retry feature as plugin. In order to create such a plugin, it needs the information about the last login attempt. The last attempt is added to the `data['conn']`.

We also face the issue, that we only call `storage_connect` before the try to connect. That's why we need a way to stop instantly when no new attempt should be done.

(followup of #7402, #5393)

hefee committed:
Add reconnect plugin to retry to login into IMAP server.

The plugin that will try to reconnect to an IMAP server, if there is no
explicit error code replied. If there is a know failure like wrong
password, no additional attempts are triggered. This should
help in cases, when the connection to the IMAP server is not 100% stable.

alecpl commented on plugins/retrylogin/readme.md:
```diff
@@ -0,0 +1,10 @@
+You can specify the maximum attempts to connect the IMAP server.
+
+  // Maximum attempts to connect the IMAP server
+  $config['retrylogin_max_attempts'] = 5;
```
Wouldn't be better to specify this as a time interval?

hefee replied:
You mean define X attempts per Y mins or you want to define the sleep between the attempts?

alecpl replied:
I meant to define maximum time from the request start to the moment you give up. I think it might be more useful that a number of attempts, but maybe it does not really matter.

hefee replied:
The current bugs #7402, #5393 both around of flaky connection to IMAP server, where attempts seems the better solution. Maybe the word timeout in title is misleading. It also handles the case, when the IMAP server closes with ""Empty startup greeting". A timeout setting would mean that the plugin needs to store the the timestamp of each first attempt to calculate the timeout, this makes the plugin more complex and we need to store this additional data.

alecpl commented on plugins/retrylogin/composer.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "retrylogin",
```
I'm not sure I like the plugin name. Maybe `imap_reconnect` or `imap_retry` would be better.

alecpl commented on plugins/retrylogin/LICENSE:
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
We don't need the LICENSE file here.

alecpl commented on program/lib/Roundcube/rcube_imap.php:
```diff
@@ -152,11 +152,16 @@ public function connect($host, $user, $pass, $port = 143, $use_ssl = null)
                 'user'           => $user,
                 'attempt'        => ++$attempt,
                 'socket_options' => [],
+                'conn'           => $this->conn,
```
$this->conn is a public property so you don't really need this.

hefee replied:
But how do the plugin get access to the current instance of rcube_imap?

alecpl replied:
rcube::get_storage()

hefee replied:
It needs to be `rcmail::get_instance()->get_storage()` at least this this line I can find in other plugins.

alecpl commented on plugins/retrylogin/retrylogin.php:
```diff
@@ -0,0 +1,52 @@
+    $args['retry'] = ($args['attempt'] <= $this->max_attempts);
+
+    if ($args['attempt'] == 1) {
+        rcube::write_log('imap','Connecting to IMAP server attempt:' . $args['attempt']);
```
We write to imap log only in debug mode. And in this mode we already have "Connecting to..." in the log. So, I would just remove this.

hefee replied:
I use roundcube 1.3.15 in production and there I do not see any "Connecting to..." line in imap log, nor can I see that this line is somewhere created in `program/lib/Roundcube/rcube_imap.php`. But things may change deeper in the stack.

I cannot really see in the communication with the server and I think it is good to know that the retry plugin is doing further attempts to login:

```
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> Connecting to IMAP server attempt:1
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: * OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE LITERAL+ STARTTLS LOGINDISABLED] Dovecot (Debian) ready.
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] C: A0001 STARTTLS
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: A0001 OK Begin TLS negotiation now.
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] C: A0002 CAPABILITY
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: * CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE LITERAL+ AUTH=PLAIN AUTH=LOGIN
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: A0002 OK Pre-login capabilities listed, post-login capabilities have more.
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] C: A0003 ID ("name" "Roundcube" "version" "1.3.15" "php" "7.3.19-1~deb10u1" "os" "Linux" "command" "/?_task=login")
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: * ID ("name" "Dovecot")
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] S: A0003 OK ID completed.
[23-Jan-2021 03:11:01 +0100]: <goet9ij9> [5055] C: A0004 AUTHENTICATE PLAIN ****** [37]
[23-Jan-2021 03:11:03 +0100]: <goet9ij9> [5055] S: A0004 NO [AUTHENTICATIONFAILED] Authentication failed.
[23-Jan-2021 03:11:03 +0100]: <goet9ij9> Connecting to IMAP server attempt:2
[...]
[23-Jan-2021 03:11:09 +0100]: <goet9ij9> Connecting to IMAP server attempt:3
```

alecpl replied:
https://github.com/roundcube/roundcubemail/blob/1eebfd723bc6706d76c280dbeaf708d4ba8f010d/program/lib/Roundcube/rcube_imap_generic.php#L1043 If you really think it is useful in non-debug mode, make it a config option in the plugin.

hefee replied:
Okay I think this is enough.

alecpl commented on plugins/retrylogin/retrylogin.php:
```diff
@@ -0,0 +1,52 @@
+    if ($args['retry']) {
+        // if we do a new attempt, sleep 50 to 150ms before retry.
+        usleep(rand(50*1000, 150*1000));
+        rcube::write_log('imap','Connecting to IMAP server attempt:' . $args['attempt']);
```
Remove this write_log.

alecpl commented on program/lib/Roundcube/rcube_imap.php:
```diff
@@ -152,11 +152,16 @@ public function connect($host, $user, $pass, $port = 143, $use_ssl = null)
 
             $data = $this->plugins->exec_hook('storage_connect', array_merge($this->options, $data));
 
+            if ($attempt > 1 && !data['retry']) {
```
You don't need this `if` statement. We check `retry` flag in the while statement below. You just set it appropriately from the plugin.

hefee replied:
here is an example without this if statement:
*attempt=1, retry= True
  -> do a `$this->conn->connect`
  -> while loop decide to go on
*attempt=2 checks the last attempt and decides not to continue retry=False
  -> still do a  `$this->conn->connect`
  -> while loop stops

as you see without this if statement another reconnect attempt is executed, but this is not what we want to do. The problem we face is that we call the storage_connect hook only before we connect and not after it. This if statement makes it possible to stop immediately and not do an unexpected reconnect attempt.

I know it is a behavior change, so we could also use another key in the array to tell to stop immediately.

alecpl replied:
I see, but I still don't like it. I'll consider this. BTW, there's missing `$` character.

hefee replied:
Fixed the `$` character.

I can fully understand that it does not looks not very clean. If you have any other idea...

alecpl commented on plugins/imap_reconnect/imap_reconnect.php:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    $storage = rcmail::get_instance()->get_storage();
+    switch ($storage->conn->errornum) {
```
You could also use more official api: $storage->get_error_code().

hefee replied:
done.

alecpl commented on plugins/imap_reconnect/config.inc.php.dist:
```diff
@@ -0,0 +1,6 @@
+// Maximum attempts to connect the IMAP server
+$config['imap_reconnect_max_attempts'] = 5;
+
+?>
```
No closing tags, please.

hefee replied:
done.

alecpl commented on plugins/imap_reconnect/composer.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "imap reconnect",
```
4 spaces indent, please.

alecpl commented on plugins/imap_reconnect/imap_reconnect.php:
```diff
@@ -0,0 +1,51 @@
+ */
+class imapreconnect extends rcube_plugin
+{
+  private $max_attempts;
```
4 spaces indent, please.

alecpl commented:
Now I have a question. Why would IMAP connection be unstable and not others. Shouldn't we implement the same for at least SMTP? What about managesieve?

Maybe the plugin should be named just 'reconnect', so we can extend it with code for other connection types in future, if needed.

hefee commented:
> Now I have a question. Why would IMAP connection be unstable and not others. Shouldn't we implement the same for at least SMTP? What about managesieve?

Well I havn't seen the root of issue for quite a while, but when we debugged the issue, we didn't found out why the IMAP connection were closed by the IMAP server only that ~1% failing and retry solved the issue. I think it may happen also for SMTP and managesieve, but as the amount of connections to SMTP and managesieve are several of magnitudes lower than IMAP, it is properly harder to see the issue.

> Maybe the plugin should be named just 'reconnect', so we can extend it with code for other connection types in future, if needed.

ACK. I understood you correctly, that I do not need to invenst time to add these connection types?

hefee committed:
Make storage_connect hook ready to be usable for reconnect plugins.

In order to decide, if we want to retry, we need the result of the last
attempt. This information is stored in $this->conn. If a plugin
decides to stop we should do not retry another login.
